lib/fog/openstack/core.rb

```diff
@@ -169,13 +173,113 @@ def self.authenticate_v2(options, connection_options = {})
       }
     end
 
+    # Keystone Style Auth
+    def self.authenticate_v3(options, connection_options = {})
+      uri                   = options[:openstack_auth_uri]
+      tenant_name           = options[:openstack_tenant]
+      service_type          = options[:openstack_service_type]
+      service_name          = options[:openstack_service_name]
+      identity_service_type = options[:openstack_identity_service_type]
+      endpoint_type         = (options[:openstack_endpoint_type] || 'public').to_s
+      openstack_region      = options[:openstack_region]
+      domain_name           = options[:openstack_domain]
```

Comment:
Have you tried using it like this????

Fog::Compute.new({:provider => 'openstack', :openstack_tenant => 'tenant_id_here', :openstack_api_key => 'tenant_password', :openstack_username => 'myusername', :openstack_auth_url => 'https://keystone/v3/auth/tokens', :openstack_domain => 'Your_domain'})

retrieve_tokens_v3 takes the openstack_username, openstack_tenant and openstack_domain and gets you the token, then authenticate_v3 makes sure that your user is inside of the token by using that information from there.